src/execution/sql/runtime_types.cpp

```diff
@@ -1,118 +1,636 @@
+  StripTime(value_, &date, &time);
+
+  // Extract minute from time.
+  int32_t hour, min, sec, fsec;
+  SplitTime(time, &hour, &min, &sec, &fsec);
+  return min;
+}
+
+int32_t Timestamp::ExtractSecond() const {
+  // Extract date component.
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  // Extract second from time.
+  int32_t hour, min, sec, fsec;
+  SplitTime(time, &hour, &min, &sec, &fsec);
+  return sec;
+}
+
+int32_t Timestamp::ExtractMillis() const {
+  // Extract date component.
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  // Extract millisecond from time.
+  int32_t hour, min, sec, fsec;
+  SplitTime(time, &hour, &min, &sec, &fsec);
+  return sec * 1000.0 + fsec / 1000.0;
+}
+
+int32_t Timestamp::ExtractMicros() const {
+  // Extract date component.
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  // Extract microsecond from time.
+  int32_t hour, min, sec, fsec;
+  SplitTime(time, &hour, &min, &sec, &fsec);
+  return sec + fsec / 1000000.0;
 }
 
-bool Date::IsValidDate(int32_t year, uint32_t month, uint32_t day) { return IsValidJulianDate(year, month, day); }
+int32_t Timestamp::ExtractDayOfWeek() const {
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  date += 1;
+  date %= 7;
+  if (date < 0) date += 7;
+  return date;
+}
+
+int32_t Timestamp::ExtractDayOfYear() const {
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  // Split date components.
+  int32_t year, month, day;
+  SplitJulianDate(date, &year, &month, &day);
+
+  // Compute date of year.
+  return BuildJulianDate(year, month, day) - BuildJulianDate(year, 1, 1) + 1;
+}
+
+void Timestamp::ExtractComponents(int32_t *year, int32_t *month, int32_t *day, int32_t *hour, int32_t *min,
+                                  int32_t *sec, int32_t *fsec) const {
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  SplitJulianDate(date, year, month, day);
+  SplitTime(time, hour, min, sec, fsec);
+}
+
+uint64_t Timestamp::ToNative() const { return value_; }
+
+Timestamp Timestamp::FromNative(Timestamp::NativeType val) { return Timestamp{val}; }
 
 std::string Timestamp::ToString() const {
-  return terrier::util::TimeConvertor::FormatTimestamp(type::timestamp_t{value_});
+  int64_t date, time;
+  StripTime(value_, &date, &time);
+
+  int32_t year, month, day, hour, min, sec, fsec;
+  ExtractComponents(&year, &month, &day, &hour, &min, &sec, &fsec);
+
+  return fmt::format("{}-{:02}-{:02} {:02}:{:02}:{:02}.{:06}", year, month, day, hour, min, sec, fsec);
 }
 
-uint64_t Timestamp::ToNative() const noexcept { return value_; }
+std::pair<bool, Timestamp> Timestamp::FromString(const char *str, std::size_t len) {
+  const char *ptr = str, *limit = ptr + len;
+
+  // Trim leading and trailing whitespace
+  while (ptr != limit && static_cast<bool>(std::isspace(*ptr))) ptr++;
+  while (ptr != limit && static_cast<bool>(std::isspace(*(limit - 1)))) limit--;
+
+  std::string suffix_str("::timestamp");
```

Comment:
Does it work to make this a `constexpr std::string_view` to avoid the object allocation every time?